proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,747 @@
+
+Mixing non-linear and linear lists (*e.g.* with ``(++)``) would
+require either some subtyping from ``List 'U a`` to ``List '1 a`` (but
+as discussed above, subptyping in presence of polymorphism quickly
+becomes hairy) or some conversion function.
+
+It it worth taking into account that the issues with ``MArray`` and
+``Array`` (which may be ``Array '1`` and ``Array 'U`` in this case)
+above are not solved by such a situation. Unless there is a subptyping
+relation from ``Array 'U`` from ``Array '1``, which cannot be performed
+by an explicit function since this would be equivalent to the
+proposal's situation.
+
+On the other hand, the CPS interface to ``newMArray`` delimits a scope
+in which the array lives. This gives a perfect opportunity to put
+clean-up code to react to exceptions. So it may not be such a bad thing
+after all.
+
+So linearity in kind seem to add a lot of complication for very little
+gain.
+
+On the matter of dependent Haskell, to the best our knowledge, the only
+presentations of dependent types with linearity-in-kinds disallow
+linear types as arguments of dependent functions.
+
+Future work
+~~~~~~~~~~~
+
+Something that hasn't been touched up by this proposal is the idea of
+declaring toplevel linear binders
+
+::
+
+  module Foo where
+  token ::('1) A  -- made up syntax
+
+Here ``token`` would have be consumed exactly once by the program,
+this property is a link-time property. This generalised the
+``RealWorld`` token which is currently magically inserted in the
+``main`` function (the existence of which is checked at link time).
+
+This would allow libraries to abstract on ``main`` or to provide their
+own linearly-threaded token.
+
+Unresolved questions
+--------------------
+
+Inference
+~~~~~~~~~
+
+- There is no systematic account of type inference. Can it be made
+  predictable when a type annotation is required? For compatibility
+  reasons, we want to infer unrestricted arrows conservatively, but
+  experience shows that it can result in very surprising type errors.
+
+- In the formalism, case expressions are indexed by a multiplicity:
+  ``case_p`` (and similarly ``let_p``). In the surface language, we
+  can deduce the multiplicity in equations when their is a type
+  annotation.
+
+  ::
+
+    fst :: (a,b) -> a
+    fst (a,_) = a    -- this is inferred as a case_ω
+
+    swap :: (a,b) ->. (b,a)
+    swap (a,b) = (b,a)   -- this is inferred as a case_1
+
+  But what of explicit ``case`` and ``let`` in the surface language? We
+  can annotate them with a multiplicity, but it is generally clear from
+  the context which multiplicity is meant. So the multiplicity
+  annotation really ought to be inferred. The general idea is: if
+  their is any linear variable in the scrutiny, then the case must be
+  linear, and if there are only unrestricted variables, it can be
+  unrestricted. Is it sound to always pick the highest possible value ?
+  What if there are multiplicities with variable multiplicity ?
```

Comment:
Looking at the scrutinee is like inferring the type of `f` in `f t` by looking at `t`. It doesn't sound right to me. We have to look at the body of `f` instead, which for a `case` it would mean the alternatives.

What's the multiplicity of the pattern variables in the alternatives?
* Are all the usages of pattern variables correct (as defined earlier in this proposal) in the alternatives? If so, the `case` can have multiplicity `One`, so it fits in more contexts.
* Is there any pattern variable with an incorrect usage? If so, the case needs to have multiplicity `Omega`, or the expression won't typecheck.